Sink linker directive computation into IRGen

hamishknight opened the pull request:
With an inverted pipeline, IRGen needs to be able to compute the linker directives itself, so sink it down such that it can be computed by the `IRGenDescriptor`.

hamishknight commented:
https://github.com/apple/llvm-project/pull/1447

@swift-ci please test

nkcsgexi reviewed (approved):
LGTM!

CodaFi commented on lib/IRGen/GenDecl.cpp:
```diff
@@ -1050,10 +1051,8 @@ void IRGenerator::emitGlobalTopLevel(llvm::StringSet<> *linkerDirectives) {
-      createLinkerDirectiveVariable(*PrimaryIGM, entry.getKey());
-    }
+  for (auto &entry: linkerDirectives) {
+    createLinkerDirectiveVariable(*PrimaryIGM, entry.getKey());
```
I know this isn't your bug, but I'm shocked we iterate over a `StringSet` here without sorting it. This cannot be correct in general.

CodaFi replied:
Also, I'm pretty sure you don't need to take the `StringRef`s here by reference since they're trivially copyable.

hamishknight replied:
> I know this isn't your bug, but I'm shocked we iterate over a StringSet here without sorting it. This cannot be correct in general.

Ouch, good spot! Looks like we also do this in `writeLdAddCFileIfNeeded`. @nkcsgexi, should TBDGen return a vector of symbols instead of a set?

hamishknight replied:
> Also, I'm pretty sure you don't need to take the StringRefs here by reference since they're trivially copyable.

It's a (`StringRef`, string length) pair, but yeah I guess `&` is a little overkill

nkcsgexi replied:
@CodaFi is deterministics your concern here or do you have other concerns?

CodaFi replied:
Yes. This code builds globals into the generated LLVM module.

hamishknight replied:
Switched to returning a vector, how does this look?

CodaFi commented on include/swift/TBDGen/TBDGen.h:
```diff
@@ -88,10 +89,7 @@ struct TBDGenOptions {
-                            const TBDGenOptions &opts);
-void enumeratePublicSymbols(ModuleDecl *module, llvm::StringSet<> &symbols,
-                            const TBDGenOptions &opts);
+llvm::StringSet<> getPublicSymbols(TBDGenDescriptor desc);
```
`TBDGenDescriptor &&` please

hamishknight replied:
Why?

CodaFi replied:
This is a consuming use of the descriptor.

hamishknight replied:
Wouldn't it be valid to re-use a descriptor? e.g make a call to get the public symbols, make a call to emit the TBD file

CodaFi replied:
Do we have usages like that? I mean, you raise a good point so this is definitely just a minor nit you can ignore if you want.

hamishknight replied:
Not that I'm aware of, though it seems like a reasonable thing to want to do, so I think I'll leave this as-is for now.

CodaFi commented on lib/FrontendTool/FrontendTool.cpp:
```diff
@@ -1781,18 +1767,13 @@ static bool performCompileStepsPostSILGen(CompilerInstance &Instance,
-  llvm::StringSet<> LinkerDirectives;
-  collectLinkerDirectives(Invocation, MSF, LinkerDirectives);
-  // Don't proceed to IRGen if collecting linker directives failed.
-  if (Context.hadError())
```
I take it we don't have tests for this failure path. We should get this to fire with a malformed TBD file.

hamishknight replied:
We have a failure path in IRGen that bails if an error is raised during emission, which should catch this case. We may do a bit of extra unnecessary IRGen compared to this early return, but I didn't think the TBDGen error path would be one that's hit all that often.

CodaFi replied:
Yeah, really only mastering errors would cause something like this.

hamishknight replied:
Looks like the test `original-defined-attr-linker-directives-fail.swift` exercises the error path, so marking this resolved.

hamishknight committed:
Have performParallelIRGeneration take an IRGenDescriptor

hamishknight committed:
[TBDGen] Make enumeratePublicSymbols more functional

Instead of taking an out parameter, have it return
the set directly. Also coalesce the two overloads
into a single overload that takes a
`TBDGenDescriptor`.

hamishknight committed:
[TBDGen] Return a vector of symbols instead of set

A couple of clients are iterating over the result,
so switch to a vector to ensure we don't
accidentally introduce any non-determinism.

hamishknight committed:
Sink linker directive computation into IRGen

With an inverted pipeline, IRGen needs to be able
to compute the linker directives itself, so sink
it down such that it can be computed by the
`IRGenDescriptor`.

hamishknight commented:
https://github.com/apple/llvm-project/pull/1447

@swift-ci please test
